### PR TITLE
Add local cache for What's New - feature announcements

### LINKED ISSUE #14585: Add local cache for What's New
Add a local cache for feature announcements to be used until the user updates to a new version.

### PR DESCRIPTION
**Pull request by Gio2018**
Fixes #14585

WordPressKit PR: https://github.com/wordpress-mobile/WordPressKit-iOS/pull/279

To test:

- put two breakpoints at lines 18 and 21 in `AnnouncementsCache.swift`
- go to Me/App Settings/What's New in WordPress
- Make sure that the execution pauses at line 21
- Resume the execution and make sure that the screen looks like the examples below
- do it again (except the breakpoints, that are already there)
- Make sure this time the execution pauses at line 18
- While you're paused, take a loot at the contents of `UserDefaults.announcements` and make sure they look consistent with the screenshots above
- Resume the execution and make sure that the screen has not changed

To reset the `UserDefaults`, pause the debugger and type

`e -l swift -- import WordPress; UserDefaults.standard.announcements = nil`

iPhone | iPad
--- | ---
 | 

PR submission checklist:

- [x] ~~I have considered adding unit tests where possible.~~ Future PR.
- [x] ~~I have considered adding accessibility improvements for my changes.~~ Future PR.
- [x] ~~I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.~~ Not released.

### CODE REVIEW

File: WordPress/Classes/ViewRelated/What's New/Data store/AnnouncementsStore.swift [hunk 1]
```diff
@@ -45,16 +48,26 @@ class RemoteAnnouncementsStore: AnnouncementsStore {
...
 
+    init(cache: AnnouncementsCache) {
+        self.cache = cache
+    }
+
     func getAnnouncements() {
-        let service = AnnouncementServiceRemote(wordPressComRestApi: api)
         state = .loading
+        if let announcements = cache.announcements {
+            self.state = .ready(announcements)
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/What's New/Data store/AnnouncementsStore.swift]: Nitpick, but don't need `self` here.

File: WordPress/Classes/ViewRelated/What's New/Data store/AnnouncementsStore.swift [hunk 2]
```diff
@@ -45,16 +48,26 @@ class RemoteAnnouncementsStore: AnnouncementsStore {
...
+
+        let service = AnnouncementServiceRemote(wordPressComRestApi: api)
         service.getAnnouncements(appId: Identifiers.appId,
                                  appVersion: Identifiers.appVersion,
                                  locale: Locale.current.identifier) { result in
 
             switch result {
             case .success(let announcements):
                 self.state = .ready(announcements)
+                self.cache.announcements = announcements
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/What's New/Data store/AnnouncementsStore.swift]: Should these two statements be the other way round? Setting the state emits the change event, but so the cached announcements should probably be in place before that happens.

File: WordPress/Classes/ViewRelated/What's New/Data store/Cache/AnnouncementsCache.swift [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            }
+            return announcements
+        }
+        set {
+            UserDefaults.standard.announcements = newValue
+        }
+    }
+
+    private func versionIsValid(for announcements: [Announcement]) -> Bool {
+        guard let minimumVersion = announcements.first?.minimumAppVersion, // there should not be more than one announcement
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/What's New/Data store/Cache/AnnouncementsCache.swift:26]: It might be good to have a couple of tests for this code.

**Review by frosty — COMMENTED** [review]
Works for me – just a couple of comments!

File: WordPress/Classes/ViewRelated/What's New/Data store/AnnouncementsStore.swift (hunk 2 above)
Comment (Gio2018) [reply WordPress/Classes/ViewRelated/What's New/Data store/AnnouncementsStore.swift]: So what sets the state is only the `getAnnouncements` method.

The flow should be:

- if there are cached values, use them and return
- if not:
  - fetch them from the backend and, as soon as you have them, use them (do not wait to save them on disk)
  - "at the same time", save them on disk, but don't let this operation block your process

So I believe that the change here would be to dispatch the save-on-disk asynchronously on the global queue

File: WordPress/Classes/ViewRelated/What's New/Data store/Cache/AnnouncementsCache.swift (hunk 3 above)
Comment (Gio2018) [reply WordPress/Classes/ViewRelated/What's New/Data store/Cache/AnnouncementsCache.swift:26]: Sure, as the description says, Unit Tests will be added in a separate PR.

**Review by mindgraffiti — APPROVED** [review]
"It's clobbering time!" 😁 

:shipit:

File: WordPress/Classes/ViewRelated/What's New/Data store/Cache/AnnouncementsCache.swift (hunk 3 above)
Comment (frosty) [reply WordPress/Classes/ViewRelated/What's New/Data store/Cache/AnnouncementsCache.swift:26]: Sorry, missed that! :)

**Pull request merged**